public/app/features/alerting/unified/components/rule-editor/LabelsField.tsx

```diff
@@ -1,17 +1,65 @@
 import { css, cx } from '@emotion/css';
-import React, { FC, useMemo } from 'react';
-import { FieldArrayWithId, useFieldArray, useFormContext } from 'react-hook-form';
+import { flattenDeep, compact, uniq } from 'lodash';
+import React, { FC, useCallback, useEffect, useMemo, useState } from 'react';
+import { useFieldArray, useFormContext } from 'react-hook-form';
 
 import { GrafanaTheme, SelectableValue } from '@grafana/data';
-import { Button, Field, InlineLabel, Label, useStyles } from '@grafana/ui';
+import { Stack } from '@grafana/experimental';
+import { Button, Field, InlineLabel, Label, useStyles, Tooltip, Icon } from '@grafana/ui';
+import { useDispatch } from 'app/types';
+import { RulerRuleGroupDTO } from 'app/types/unified-alerting-dto';
 
+import { useUnifiedAlertingSelector } from '../../hooks/useUnifiedAlertingSelector';
+import { fetchRulerRulesIfNotFetchedYet } from '../../state/actions';
 import { RuleFormValues } from '../../types/rule-form';
-import AlertLabelDropdown, { AlertLabelDropdownProps } from '../AlertLabelDropdown';
+import { GRAFANA_RULES_SOURCE_NAME } from '../../utils/datasource';
+import AlertLabelDropdown from '../AlertLabelDropdown';
 
 interface Props {
   className?: string;
 }
 
+const useGetCustomLabels = () => {
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    dispatch(fetchRulerRulesIfNotFetchedYet(GRAFANA_RULES_SOURCE_NAME));
+  }, [dispatch]);
+
+  const rulerRuleRequests = useUnifiedAlertingSelector((state) => state.rulerRules);
+
+  const rulerRequest = rulerRuleRequests[GRAFANA_RULES_SOURCE_NAME];
+
+  if (!rulerRequest || rulerRequest.loading) {
+    return;
+  }
+
+  const result = rulerRequest.result || {};
+
+  //store all labels in a flat array and remove empty values
+  const labels = compact(
+    flattenDeep(
+      Object.keys(result).map((ruleGroupKey) =>
+        result[ruleGroupKey].map((ruleItem: RulerRuleGroupDTO) => ruleItem.rules.map((item) => item.labels))
+      )
+    )
+  );
+
+  const labelsByKey: Record<string, string[]> = {};
+
+  labels.forEach((label: Record<string, string>) => {
+    Object.entries(label).forEach(([key, value]) => {
+      labelsByKey[key] = uniq([...(labelsByKey[key] || []), value]);
```

Comment:
nit: How about using `Set` to keep labels unique?

Reply:
Good idea 👍